tests/components/hassmpris/test_config_flow.py

```diff
@@ -0,0 +1,347 @@
+        """Initialize the mock."""
+        pass
+
+    async def ping(self):
+        """Fake successful ping."""
+
+
+class fakecsrkey:
+    """Creates CSR and key request, caching the same one between calls."""
+
+    def __init__(self):
+        """Initialize the faker."""
+        self.csr, self.key = pskca.create_certificate_signing_request()
+
+    def __call__(self):
+        """Return the CSR and key when the patch() patches pskca with this."""
+        return self.csr, self.key
+
+
+_hostinfo = {
+    "host": "1.1.1.1",
+    "cakes_port": 40052,
+    "mpris_port": 40051,
+}
+
+_zeroconfinfo = zeroconf.ZeroconfServiceInfo(
+    host="127.0.0.1",
+    addresses=[],
+    port=40051,
+    hostname="uniqueid",
+    name="thename",
+    properties={CONF_CAKES_PORT: "40052"},
+    type="_hassmpris._tcp.local.",
+)
+
+
+async def test_user_flow(hass: HomeAssistant) -> None:
+    """Test we get the user form and, upon success, go to confirm step."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == FlowResultType.FORM
+    assert not result["errors"]
+
+    with patch(
+        "hassmpris_client.AsyncCAKESClient", return_value=MockCakesClient()
+    ), patch("pskca.create_certificate_signing_request", fakecsrkey()):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            _hostinfo,
+        )
+        await hass.async_block_till_done()
+
+        assert result2["type"] == FlowResultType.FORM
+        assert result2["step_id"] == STEP_CONFIRM
+
+        result3 = await hass.config_entries.flow.async_configure(
+            result2["flow_id"],
+            {
+                "emojis": "doesn't matter",
+            },
+        )
+        await hass.async_block_till_done()
+
+        assert result3["type"] == FlowResultType.CREATE_ENTRY
+
+    # Now we test that the entry is not actually added twice.
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        _hostinfo,
+    )
+    await hass.async_block_till_done()
+    assert result["type"] == FlowResultType.ABORT
+
+
+async def test_user_flow_errors(hass: HomeAssistant) -> None:
+    """Test we get the user form and then fails."""
+    with patch("pskca.create_certificate_signing_request", fakecsrkey()):
+        for exc, reason in [
+            (hassmpris_client.ClientException, REASON_CANNOT_CONNECT),
+            (hassmpris_client.Timeout, REASON_TIMEOUT),
+        ]:
+            # test the connection error.
+            result = await hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": config_entries.SOURCE_USER}
+            )
+            with patch(
+                "hassmpris_client.AsyncCAKESClient",
```

Comment:
Please patch the library at the import port instead

Reply:
That cannot be done.  Tests need different behaviors from the mock nearly every time this is patched.  Hence the patch decorators when appropriate.

If you have another suggestion, I will gladly reconsider.  I will leave this item open in the review for you to reply.